src/ocl_utils_mod.f90

```diff
@@ -121,65 +121,88 @@ function get_program(context, device, version_str, filename) result(prog)
     integer(c_intptr_t), intent(inout), target :: device, context 
     character(len=CL_UTIL_STR_LEN), intent(in) :: version_str, filename
     ! Locals
-    character(len=1,kind=c_char), allocatable, target :: source(:)
+    character(len=1, kind=c_char), allocatable, target :: source(:)
     type(c_ptr), target :: psource
     character(len=1024) :: options
+    character(len=3) :: extension
     character(len=1, kind=c_char), target :: retinfo(1:1024), c_options(1:1024)
-    integer :: i, irec, iallocerr
+    integer :: i, irec, iallocerr, last_dot
     integer(c_int32_t) :: ierr
     integer, parameter :: iunit=10
     integer(c_size_t), target :: binary_size, iret
     character, dimension(1) :: char
 
+    ! Get the filename extension
+    last_dot = scan(trim(filename), ".", BACK= .true.)
+    extension = filename(last_dot:)
+
     ! read kernel from disk
-    open(iunit, file=filename, access='direct', &
-         status='old', action='read', iostat=ierr, recl=1)
-    if (ierr.ne.0)then
+    open(iunit, file=filename, access='direct', status='old', action='read', &
+         iostat=ierr, recl=1)
+    if (ierr.ne.0) then
        write(*,*) 'Cannot open file: ', TRIM(filename)
        stop
     end if
     irec=1
     do
        read(iunit, rec=irec, iostat=ierr) char
-       if (ierr /= 0) exit
+       if (ierr.ne.0) exit
        irec = irec+1
     end do
-
     if (irec.eq.0) stop 'nothing read'
-    allocate(source(irec+1),stat=iallocerr)
+    allocate(source(irec+1), stat=iallocerr)
     if (iallocerr.ne.0) stop 'allocate'
     do i=1,irec
        read(iunit,rec=i,iostat=ierr) source(i:i)
     enddo
     close(iunit)
-
-    print '(a,i7)','size of source code in bytes: ',irec
-
     psource=C_LOC(source) ! pointer to source code
-    binary_size = irec
-    prog = clCreateProgramWithBinary(context, 1, C_LOC(device), &
-                                     C_LOC(binary_size), C_LOC(psource), &
-                                     C_NULL_PTR, ierr)
 
-    call check_status('clCreateProgramWithSource', ierr)
+    print '(a,i7)', 'size of source code in bytes: ', irec
+
+    ! Create the OpenCL program
+    if (extension == ".cl") then
+        ! If it has a .cl extension, it will be JIT'ed
+        source(irec+1) = C_NULL_CHAR  ! In C strings end with NULL 
+        binary_size = irec + 1
+        prog = clCreateProgramWithSource(context, 1, C_LOC(psource), &
+                                         C_LOC(binary_size), ierr)
+        call check_status('clCreateProgramWithSource', ierr)
+    else
```

Comment:
I know I didn't write one originally but please could you add a brief 'docstring' for this routine saying what it does.

Reply:
Done